plugin/restore/mysql/mysql.go

```diff
@@ -83,6 +86,84 @@ func (r *Restore) RestorePITR(ctx context.Context, fullBackup *bufio.Scanner, bi
 	return nil
 }
 
+// Parse the binlog at position, and return the timestamp in the binlog event.
+// The current mechanism is by invoking mysqlbinlog and parse the output string.
+// Maybe we should parse the raw binlog header to get better documented structure?
+// nolint
+func (r *Restore) parseBinlogEventTimestamp(ctx context.Context, binlogInfo api.BinlogInfo, binlogDir string) (int64, error) {
+	args := []string{
+		path.Join(binlogDir, binlogInfo.FileName),
+		fmt.Sprintf("--start-position %d", binlogInfo.Position),
+		// This will trick mysqlbinlog to output the binlog event header followed by a warning message telling that
+		// the --stop-position is in the middle of the binlog event.
+		// It's OK, since we are only parsing for the timestamp in the binlog event header.
+		fmt.Sprintf("--stop-position %d", binlogInfo.Position+1),
+	}
+	var buf bytes.Buffer
+	cmd := exec.CommandContext(ctx, r.mysqlbinlog.GetPath(), args...)
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = &buf
+
+	if err := cmd.Run(); err != nil {
+		return -1, err
+	}
+
+	timestamp, err := parseBinlogEventTimestampImpl(buf.String())
+	if err != nil {
+		return timestamp, fmt.Errorf("failed to parse binlog event timestamp, filename[%s], position[%d], error[%w]", binlogInfo.FileName, binlogInfo.Position, err)
+	}
+
+	return timestamp, nil
+}
+
+func parseBinlogEventTimestampImpl(output string) (int64, error) {
+	lines := strings.Split(output, "\n")
+	// The mysqlbinlog output will contains a line starting with "#220421 14:49:26 server id 1",
+	// which has the timestamp we are looking for.
+	// The first occurrence is the target.
+	for _, line := range lines {
+		if strings.Contains(line, "server id") {
+			date, err := time.Parse("060102 15:04:05", line[1:16])
+			if err != nil {
+				return -1, err
+			}
+			return date.Unix(), nil
+		}
+	}
+	return -1, fmt.Errorf("no timestamp found in mysqlbinlog output")
+}
+
+// Find the latest logical backup and corresponding binlog info whose time is before `restoreTs`.
+// The backupList should only contain DONE backups, and sorted that the newest backup is at the front.
+// TODO(dragonly)/TODO(zp): Use this when the apply binlog PR is ready, and remove the nolint comments
+// nolint
+func (r *Restore) getNewestBackupAfterTs(ctx context.Context, backupList []*api.Backup, restoreTs int64, binlogDir string) (*api.Backup, error) {
+	backup, err := r.getNewestBackupAfterTsImpl(ctx, backupList, restoreTs, binlogDir)
+	if err != nil {
+		return nil, err
+	}
+
+	return backup, nil
+}
+
+// Traverse the backups starting from the latest and find the first with timestamp less than the target restore timestamp.
+// nolint
+func (r *Restore) getNewestBackupAfterTsImpl(ctx context.Context, backupList []*api.Backup, restoreTs int64, binlogDir string) (*api.Backup, error) {
```

Comment:
Is the function name correct?

I thought we're getting the latest backup before Ts?